core/commands/dag/dag.go

```diff
@@ -96,56 +92,93 @@ into an object of the specified format.
 			}
 		}
 
-		if dopin {
-			defer n.Blockstore.PinLock().Unlock()
-		}
+		outChan := make(chan interface{}, 8)
+		res.SetOutput((<-chan interface{})(outChan))
 
-		nds, err := coredag.ParseInputs(ienc, format, fi, mhType, -1)
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
-		}
-		if len(nds) == 0 {
-			res.SetError(fmt.Errorf("no node returned from ParseInputs"), cmds.ErrNormal)
-			return
-		}
+		addAllAndPin := func(f files.File) error {
+			for {
+				file, err := f.NextFile()
+				if err == io.EOF {
+					// Finished the list of files.
+					break
+				} else if err != nil {
+					return err
+				}
 
-		b := n.DAG.Batch()
-		for _, nd := range nds {
-			_, err := b.Add(nd)
-			if err != nil {
-				res.SetError(err, cmds.ErrNormal)
-				return
+				nds, err := coredag.ParseInputs(ienc, format, file, mhType, -1)
+				if err != nil {
+					return err
+				}
+				if len(nds) == 0 {
+					return fmt.Errorf("no node returned from ParseInputs")
+				}
+
+				var cid *cid.Cid
+				err = func() error {
+					if dopin {
+						defer n.Blockstore.PinLock().Unlock()
+					}
+
+					b := n.DAG.Batch()
+					for _, nd := range nds {
+						_, err := b.Add(nd)
+						if err != nil {
+							return err
+						}
+					}
+
+					if err := b.Commit(); err != nil {
+						return err
+					}
+
+					cid = nds[0].Cid()
+					if dopin {
+						n.Pinning.PinWithMode(cid, pin.Recursive)
```

Comment:
Also, since we're batching, we should just gather up all the things we're going to pin into an array, pin them all at once at the end, then only call Pinning.Flush once